### DISCUSSION TITLE
EXODIFF errors in MOOSE on an HPC cluster

### DISCUSSION THREAD

**Discussion by adfboyd** [idaholab/moose#18094] [Q&A General] [answered]
Hello,

I was advised to start a new discussion on the remaining errors I am experiencing in testing MOOSE on a cluster. The final test results from ./run_tests have been attached. Any input on how to fix these would be much appreciated. (the previous discussion on installation and testing can be found here: https://github.com/idaholab/moose/discussions/17946) 
[finaltestresults.txt](https://github.com/idaholab/moose/files/6662512/finaltestresults.txt)
Thank you!

**Comment by milljm** [comment]
Sorry, I don't think I was clear; The developers will want the error _before_ the 'Final Test Results'. The full error listed before the final results contains the reason why exodiff failed.

**Reply by adfboyd** [reply]
Apologies - the whole thing is here:
[slurm-331094.txt](https://github.com/idaholab/moose/files/6663708/slurm-331094.txt)

**Comment by lindsayad** [comment, marked as answer]
Most of the errors are due to missing python packages. Definitely matplotlib is missing. Is numpy available?

One exodiff is trivial (0 vs. 1e-17 ... I don't even know why that would have registered because we have a default floor of 1e-10) so I wouldn't worry about that. The contact-related exodiffs are interesting...basically the same result but different sign. I'm going to have to look into that a bit more to understand how that could be.

Then there are a couple of out-of-memory errors due to extensive mesh refinement I think.

Overall I wouldn't be too worried about these errors.

**Reply by lindsayad** [reply]
Actually this warning looks a little scarier:
```
Warning: unable to write derivative cache file.
```

**Reply by lindsayad** [reply]
@dschwen any thoughts on the above warning? Looks to be out of `fparser_ad.cc`

**Reply by lindsayad** [reply]
@adfboyd I think it would actually be helpful for me if you could send the file `moose/test/tests/geomsearch/3d_moving_penetration/pl_test4q_out.e`

**Reply by GiudGiud** [reply]
@lindsayad I ll add the python package dependencies to the tests that are not checking
Also there s an error showing cli_args mispelled. I ll look at that

**Reply by GiudGiud** [reply]
A few of the test failures are due to a typo in the diffs to analyze_jacobian.py.
It should be cli_args instead of cli_ags.
There s other failures with that test scripts, not exodiffs though

**Reply by LabrosV** [reply]
> Actually this warning looks a little scarier:
> 
> ```
> Warning: unable to write derivative cache file.
> ```

I'm getting the same warning when running my app on an HPC cluster. 
Convergence and results look fine. Any way to fix it?

**Reply by GiudGiud** [reply]
This should not influence the results according to @dschwen in https://github.com/idaholab/moose/discussions/16464

**Comment by adfboyd** [comment]
You can find the file here: https://drive.google.com/file/d/1GI8hdW2Zt3ZtjFnSjkk6UOecYkDMIY65/view?usp=sharing .
Please let me know if there's a better way of sharing this type of thing, it's not supported on Github.

**Reply by lindsayad** [reply]
I don't know what to think. The normal should always point out of the primary domain but it's pointing into the domain on your system for that single time step. Moreover, distance should be positive if the faces are penetrated (which they are here) and it's negative. That's a direct result of the wrong normal computation though. @friedmud is the original architect of the `FindContactPoint` code. Maybe he has some thoughts but I don't see anything in the code that would make me think the result should ever be wrong just because you're on a different architecture. For @friedmud or others, the TLDR is we have this:

<img width="1002" alt="Screen Shot 2021-06-16 at 11 05 53 AM" src="https://user-images.githubusercontent.com/10248304/122270158-e0620780-ce92-11eb-8721-80530085657d.png">
 
and we are getting the exodiff
```
geomsearch/3d_moving_penetration.pl_test4q: Nodal variables:
geomsearch/3d_moving_penetration.pl_test4q:    distance             rel diff:  4.2426407e-01 ~ -4.2426407e-01 = 2.00000e+00 (node 37)
geomsearch/3d_moving_penetration.pl_test4q:    normal_x             rel diff: -7.0710678e-01 ~  7.0710678e-01 = 2.00000e+00 (node 37)
geomsearch/3d_moving_penetration.pl_test4q:    normal_y             rel diff: -7.0710678e-01 ~  7.0710678e-01 = 2.00000e+00 (node 37)
```

Here's with the primary block volume so that you know which way the normal should point (it's correct in the gold file, file 1)

<img width="845" alt="Screen Shot 2021-06-16 at 11 07 55 AM" src="https://user-images.githubusercontent.com/10248304/122271048-d4c31080-ce93-11eb-9fb9-0c38fa77a3f6.png">

**Reply by GiudGiud** [reply]
@friedmud any thoughts on this?
